components/gfx/display_list/mod.rs

```diff
@@ -323,6 +323,123 @@ impl DisplayList {
         paint_subcontext.pop_clip_if_applicable();
         paint_subcontext.draw_target.set_transform(&old_transform)
     }
+
+    pub fn hit_test(&self,
+                    point: Point2D<Au>,
+                    result: &mut Vec<DisplayItemMetadata>,
+                    topmost_only: bool) {
+        fn hit_test_in_list<'a, I>(point: Point2D<Au>,
+                                   result: &mut Vec<DisplayItemMetadata>,
+                                   topmost_only: bool,
+                                   iterator: I)
+                                   where I: Iterator<Item=&'a DisplayItem> {
+            for item in iterator {
+                // TODO(pcwalton): Use a precise algorithm here. This will allow us to properly hit
+                // test elements with `border-radius`, for example.
+                if !item.base().clip.might_intersect_point(&point) {
+                    // Clipped out.
+                    continue
+                }
+                if !geometry::rect_contains_point(item.bounds(), point) {
+                    // Can't possibly hit.
+                    continue
+                }
+                if item.base().metadata.pointing.is_none() {
+                    // `pointer-events` is `none`. Ignore this item.
+                    continue
+                }
+                match *item {
+                    DisplayItem::BorderClass(ref border) => {
```

Comment:
nit: While you're here, could you change this to `if let`? That will eliminate a little bit of the excess rightward drift in the point and size code below.